File: FBDeviceControl/Management/FBDeviceSet.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceSet.h"

#import <FBControlCore/FBControlCore.h>
#import <FBControlCore/FBiOSTargetSet.h>
#import <FBControlCore/FBiOSTarget.h>

#import <objc/runtime.h>

#import "FBAMDevice+Private.h"
#import "FBAMDevice.h"
#import "FBAMDeviceManager.h"
#import "FBAMRestorableDevice.h"
#import "FBAMRestorableDeviceManager.h"
#import "FBDevice+Private.h"
#import "FBDevice.h"
#import "FBDeviceControlFrameworkLoader.h"
#import "FBDeviceStorage.h"

@interface FBDeviceSet () <FBiOSTargetSetDelegate>

@property (nonatomic, strong, readonly) FBAMDeviceManager *amDeviceManager;
@property (nonatomic, strong, readonly) FBAMRestorableDeviceManager *restorableDeviceManager;
@property (nonatomic, strong, readonly) FBDeviceStorage<FBDevice *> *storage;

@end

@implementation FBDeviceSet

@synthesize delegate = _delegate;

#pragma mark Initializers

+ (void)initialize
{
  [FBDeviceControlFrameworkLoader.new loadPrivateFrameworksOrAbort];
}

+ (nullable instancetype)setWithLogger:(id<FBControlCoreLogger>)logger delegate:(id<FBiOSTargetSetDelegate>)delegate ecidFilter:(NSString *)ecidFilter error:(NSError **)error
{
  dispatch_queue_t workQueue = dispatch_get_main_queue();
  dispatch_queue_t asyncQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0);
  return [FBDeviceSet setWithWorkQueue:workQueue asyncQueue:asyncQueue logger:logger delegate:delegate ecidFilter:ecidFilter error:error];
}

+ (nullable instancetype)setWithWorkQueue:(dispatch_queue_t)workQueue asyncQueue:(dispatch_queue_t)asyncQueue logger:(id<FBControlCoreLogger>)logger delegate:(id<FBiOSTargetSetDelegate>)delegate ecidFilter:(NSString *)ecidFilter error:(NSError **)error
{
  AMDCalls calls = FBDeviceControlFrameworkLoader.amDeviceCalls;
  FBAMDeviceManager *amDeviceManager = [[FBAMDeviceManager alloc] initWithCalls:calls workQueue:workQueue asyncQueue:asyncQueue ecidFilter:ecidFilter logger:logger];
  FBAMRestorableDeviceManager *restorableDeviceManager = [[FBAMRestorableDeviceManager alloc] initWithCalls:calls workQueue:workQueue asyncQueue:asyncQueue ecidFilter:ecidFilter logger:logger];
  FBDeviceSet *deviceSet = [[FBDeviceSet alloc] initWithAMDeviceManager:amDeviceManager restorableDeviceManager:restorableDeviceManager logger:logger delegate:delegate];
  if (![amDeviceManager startListeningWithError:error]) {
    return nil;
  }
  if (![restorableDeviceManager startListeningWithError:error]) {
    return nil;
  }
  return deviceSet;
}

- (instancetype)initWithAMDeviceManager:(FBAMDeviceManager *)amDeviceManager restorableDeviceManager:(FBAMRestorableDeviceManager *)restorableDeviceManager logger:(id<FBControlCoreLogger>)logger delegate:(id<FBiOSTargetSetDelegate>)delegate
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _amDeviceManager = amDeviceManager;
  _restorableDeviceManager = restorableDeviceManager;
  _delegate = delegate;
  _logger = logger;
  _storage = [[FBDeviceStorage alloc] initWithLogger:logger];

  [self subscribeToDeviceNotifications];

  return self;
}

- (void)dealloc
{
  [self unsubscribeFromDeviceNotifications];
}

- (NSString *)description
{
  return [NSString stringWithFormat:@"FBDeviceSet: %@", [FBCollectionInformation oneLineDescriptionFromArray:self.allDevices]];
}

#pragma mark Querying

- (id<FBiOSTargetInfo>)targetWithUDID:(NSString *)udid
{
  return [self deviceWithUDID:udid];
}

- (FBDevice *)deviceWithUDID:(NSString *)udid
{
  return [[self.allDevices filteredArrayUsingPredicate:FBiOSTargetPredicateForUDID(udid)] firstObject];
}

#pragma mark FBiOSTargetSet Implementation

- (NSArray<id<FBiOSTarget>> *)allTargetInfos
{
  return self.allDevices;
}

#pragma mark Private

- (void)subscribeToDeviceNotifications
{
  self.amDeviceManager.delegate = self;
  self.restorableDeviceManager.delegate = self;
  for (FBAMDevice *amDevice in self.amDeviceManager.currentDeviceList) {
    [self targetAdded:amDevice inTargetSet:self.amDeviceManager];
  }
  for (FBAMRestorableDevice *restorableDevice in self.restorableDeviceManager.currentDeviceList) {
    [self targetAdded:restorableDevice inTargetSet:self.restorableDeviceManager];
  }
}

- (void)unsubscribeFromDeviceNotifications
{
  self.amDeviceManager.delegate = nil;
  self.restorableDeviceManager.delegate = nil;
}

- (void)amDeviceAdded:(FBAMDevice *)amDevice
{
  FBDevice *device = [self.storage deviceForKey:amDevice.uniqueIdentifier];
  if (device) {
    device.amDevice = amDevice;
  } else {
    device = [[FBDevice alloc] initWithSet:self amDevice:amDevice restorableDevice:nil logger:self.logger];
    [self.storage deviceAttached:device forKey:amDevice.uniqueIdentifier];
  }
  [self.delegate targetAdded:device inTargetSet:self];
}

- (void)amDeviceRemoved:(FBAMDevice *)amDevice
{
  FBDevice *device = [self.storage deviceForKey:amDevice.uniqueIdentifier];
  if (!device) {
    [self.logger logFormat:@"%@ was removed, but there's no active device for it", amDevice];
    return;
  }
  device.amDevice = NULL;
  if (device.amDevice || device.restorableDevice) {
    [self.delegate targetUpdated:device inTargetSet:self];
  } else {
    [self.storage deviceDetachedForKey:amDevice.uniqueIdentifier];
    [self.delegate targetRemoved:device inTargetSet:self];
  }
}

- (void)restorableDeviceAdded:(FBAMRestorableDevice *)restorableDevice
{
  FBDevice *device = [self.storage deviceForKey:restorableDevice.uniqueIdentifier];
  if (device) {
    device.restorableDevice = restorableDevice;
  } else {
    device = [[FBDevice alloc] initWithSet:self amDevice:nil restorableDevice:restorableDevice logger:self.logger];
    [self.storage deviceAttached:device forKey:restorableDevice.uniqueIdentifier];
  }
  [self.delegate targetAdded:device inTargetSet:self];
}

- (void)restorableDeviceRemoved:(FBAMRestorableDevice *)restorableDevice
{
  FBDevice *device = [self.storage deviceForKey:restorableDevice.uniqueIdentifier];
  if (!device) {
    [self.logger logFormat:@"%@ was removed, but there's no active device for it", restorableDevice];
    return;
  }
  device.restorableDevice = NULL;
  if (device.amDevice || device.restorableDevice) {
    [self.delegate targetUpdated:device inTargetSet:self];
  } else {
    [self.storage deviceDetachedForKey:restorableDevice.uniqueIdentifier];
    [self.delegate targetRemoved:device inTargetSet:self];
  }
}

#pragma mark Properties

- (NSArray<FBDevice *> *)allDevices
{
  return [self.storage.attached.allValues sortedArrayUsingSelector:@selector(uniqueIdentifier)];
}

#pragma mark FBiOSTargetSetDelegate Implementation

- (void)targetAdded:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{
  if ([targetInfo isKindOfClass:FBAMDevice.class]) {
    [self amDeviceAdded:(FBAMDevice *) targetInfo];
  } else if ([targetInfo isKindOfClass:FBAMRestorableDevice.class]) {
    [self restorableDeviceAdded:(FBAMRestorableDevice *) targetInfo];
  } else {
    [self.logger logFormat:@"Ignoring %@ as it is not a valid target type", targetInfo];
  }
}

- (void)targetRemoved:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{
  if ([targetInfo isKindOfClass:FBAMDevice.class]) {
    [self amDeviceRemoved:(FBAMDevice *) targetInfo];
  } else if ([targetInfo isKindOfClass:FBAMRestorableDevice.class]) {
    [self restorableDeviceRemoved:(FBAMRestorableDevice *) targetInfo];
  } else {
    [self.logger logFormat:@"Ignoring %@ as it is not a valid target type", targetInfo];
  }
}

- (void)targetUpdated:(id<FBiOSTargetInfo>)targetInfo inTargetSet:(id<FBiOSTargetSet>)targetSet
{
  FBDevice *device = [self.storage deviceForKey:targetInfo.uniqueIdentifier];
  if (device && [targetInfo isKindOfClass:FBAMDevice.class]) {
    device.amDevice = (FBAMDevice *) targetInfo;
  } else if (device && [targetInfo isKindOfClass:FBAMRestorableDevice.class]) {
    device.restorableDevice = (FBAMRestorableDevice *) targetInfo;
  } else {
    NSAssert(NO, @"No existing device to update for %@", targetInfo);
  }
  [self.delegate targetUpdated:device inTargetSet:self];
}

@end

File: FBDeviceControl/Management/FBDeviceSet.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import FBControlCore
import Foundation

public final class FBDeviceSet: FBiOSTargetSet, FBiOSTargetSetDelegate, CustomStringConvertible {
  // Loaded once per process; a failed load is cached and rethrown by every init rather than
  // aborting the process.
  private static let _amDeviceCalls: Result<AMDCalls, Error> = Result {
    let loader = FBDeviceControlFrameworkLoader()
    try loader.loadPrivateFrameworks(FBControlCoreGlobalConfiguration.defaultLogger)
    return try loader.amDeviceCalls
  }

  private let amDeviceManager: FBAMDeviceManager
  private let restorableDeviceManager: FBAMRestorableDeviceManager
  private let storage: FBDeviceStorage<FBDevice>
  public let logger: any FBControlCoreLogger
  public weak var delegate: (any FBiOSTargetSetDelegate)?

  // MARK: Initializers

  public convenience init(logger: any FBControlCoreLogger, delegate: (any FBiOSTargetSetDelegate)?, ecidFilter: String?) throws {
    let calls = try Self._amDeviceCalls.get()
    let workQueue = DispatchQueue.main
    let asyncQueue = DispatchQueue.global(qos: .userInitiated)
    let amDeviceManager = FBAMDeviceManager(calls: calls, work: workQueue, asyncQueue: asyncQueue, ecidFilter: ecidFilter, logger: logger)
    let restorableDeviceManager = FBAMRestorableDeviceManager(calls: calls, work: workQueue, asyncQueue: asyncQueue, ecidFilter: ecidFilter, logger: logger)
    self.init(amDeviceManager: amDeviceManager, restorableDeviceManager: restorableDeviceManager, logger: logger, delegate: delegate)
    try amDeviceManager.startListening()
    try restorableDeviceManager.startListening()
  }

  private init(amDeviceManager: FBAMDeviceManager, restorableDeviceManager: FBAMRestorableDeviceManager, logger: any FBControlCoreLogger, delegate: (any FBiOSTargetSetDelegate)?) {
    self.amDeviceManager = amDeviceManager
    self.restorableDeviceManager = restorableDeviceManager
    self.logger = logger
    self.delegate = delegate
    self.storage = FBDeviceStorage<FBDevice>(logger: logger)
    subscribeToDeviceNotifications()
  }

  deinit {
    unsubscribeFromDeviceNotifications()
  }

  // MARK: CustomStringConvertible

  public var description: String {
    "FBDeviceSet: \(FBCollectionInformation.oneLineDescription(from: allDevices))"
  }

  // MARK: Querying

  public func target(withUDID udid: String) -> (any FBiOSTargetInfo)? {
    deviceWithUDID(udid)
  }

  public func deviceWithUDID(_ udid: String) -> FBDevice? {
    allDevices.first { $0.udid == udid }
  }

  // MARK: FBiOSTargetSet

  public var allTargetInfos: [any FBiOSTargetInfo] {
    allDevices
  }

  // MARK: Properties

  public var allDevices: [FBDevice] {
    Array(storage.attached.values).sorted { $0.uniqueIdentifier < $1.uniqueIdentifier }
  }

  // MARK: Private

  private func subscribeToDeviceNotifications() {
    amDeviceManager.delegate = self
    restorableDeviceManager.delegate = self
    for amDevice in amDeviceManager.currentDeviceList {
      targetAdded(amDevice, in: amDeviceManager)
    }
    for restorableDevice in restorableDeviceManager.currentDeviceList {
      targetAdded(restorableDevice, in: restorableDeviceManager)
    }
  }

  private func unsubscribeFromDeviceNotifications() {
    amDeviceManager.delegate = nil
    restorableDeviceManager.delegate = nil
  }

  private func amDeviceAdded(_ amDevice: FBAMDevice) {
    if let device = storage.device(forKey: amDevice.uniqueIdentifier) {
      device.amDevice = amDevice
    } else {
      let device = FBDevice(set: self, amDevice: amDevice, restorableDevice: nil, logger: logger)
      storage.deviceAttached(device, forKey: amDevice.uniqueIdentifier)
    }
    if let device = storage.device(forKey: amDevice.uniqueIdentifier) {
      delegate?.targetAdded(device, in: self)
    }
  }

  private func amDeviceRemoved(_ amDevice: FBAMDevice) {
    guard let device = storage.device(forKey: amDevice.uniqueIdentifier) else {
      logger.log("\(amDevice) was removed, but there's no active device for it")
      return
    }
    device.amDevice = nil
    if device.restorableDevice != nil {
      delegate?.targetUpdated(device, in: self)
    } else {
      storage.deviceDetached(forKey: amDevice.uniqueIdentifier)
      delegate?.targetRemoved(device, in: self)
    }
  }

  private func restorableDeviceAdded(_ restorableDevice: FBAMRestorableDevice) {
    if let device = storage.device(forKey: restorableDevice.uniqueIdentifier) {
      device.restorableDevice = restorableDevice
    } else {
      let device = FBDevice(set: self, amDevice: nil, restorableDevice: restorableDevice, logger: logger)
      storage.deviceAttached(device, forKey: restorableDevice.uniqueIdentifier)
    }
    if let device = storage.device(forKey: restorableDevice.uniqueIdentifier) {
      delegate?.targetAdded(device, in: self)
    }
  }

  private func restorableDeviceRemoved(_ restorableDevice: FBAMRestorableDevice) {
    guard let device = storage.device(forKey: restorableDevice.uniqueIdentifier) else {
      logger.log("\(restorableDevice) was removed, but there's no active device for it")
      return
    }
    device.restorableDevice = nil
    if device.amDevice != nil {
      delegate?.targetUpdated(device, in: self)
    } else {
      storage.deviceDetached(forKey: restorableDevice.uniqueIdentifier)
      delegate?.targetRemoved(device, in: self)
    }
  }

  // MARK: FBiOSTargetSetDelegate

  public func targetAdded(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {
    if let amDevice = targetInfo as? FBAMDevice {
      amDeviceAdded(amDevice)
    } else if let restorableDevice = targetInfo as? FBAMRestorableDevice {
      restorableDeviceAdded(restorableDevice)
    } else {
      logger.log("Ignoring \(targetInfo) as it is not a valid target type")
    }
  }

  public func targetRemoved(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {
    if let amDevice = targetInfo as? FBAMDevice {
      amDeviceRemoved(amDevice)
    } else if let restorableDevice = targetInfo as? FBAMRestorableDevice {
      restorableDeviceRemoved(restorableDevice)
    } else {
      logger.log("Ignoring \(targetInfo) as it is not a valid target type")
    }
  }

  public func targetUpdated(_ targetInfo: any FBiOSTargetInfo, in targetSet: any FBiOSTargetSet) {
    guard let device = storage.device(forKey: targetInfo.uniqueIdentifier) else {
      assertionFailure("No existing device to update for \(targetInfo)")
      return
    }
    if let amDevice = targetInfo as? FBAMDevice {
      device.amDevice = amDevice
    } else if let restorableDevice = targetInfo as? FBAMRestorableDevice {
      device.restorableDevice = restorableDevice
    } else {
      assertionFailure("No existing device to update for \(targetInfo)")
      return
    }
    delegate?.targetUpdated(device, in: self)
  }
}
